lib/model/channel.dart

```diff
@@ -199,6 +199,51 @@ class ChannelStoreImpl with ChannelStore {
           streamsByName.remove(stream.name);
           subscriptions.remove(stream.streamId);
         }
+
+      case ChannelUpdateEvent():
+        final ChannelUpdateEvent(:streamId, name: String streamName) = event;
+        assert(identical(streams[streamId], streamsByName[streamName]));
```

Comment:
We use `assert`s when there's an invariant we're managing purely in our own code. But here, the assert could fail because of unexpected server behavior. When we want to guard against unexpected server behavior, we don't use asserts. In production, `asserts` are ignored, and also, arguments to them aren't evaluated: https://dart.dev/language/error-handling#assert . We might instead throw an error, or just give up on the thing we were trying to do and move on, usually with a `// TODO(log)`.

Here, though, we don't need the channel name on the event, do we? We can just look it up on the `ZulipStream` instance in our data structures. It seems potentially nice to avoid depending on the channel name in the event, to give the API more freedom to deprecate/remove that field as a cleanup.

Reply:
I think the `assert` is actually for making sure our two local copies in `streams` and `streamsByName` are the same. The same assert is used in `ChannelDeleteEvent`.

`streamName` is actually useful when updating `streamsByName` when the event is about updating the channel name:
```dart
case ChannelPropertyName.name:
  channel.name = event.value as String;
  streamsByName.remove(streamName); 
  streamsByName[channel.name] = channel;
```
But if we really want to prevent getting `streamName` from the event, we can get it by storing `channel.name` in a temporary variable. Should I go with the latter option?

Reply:
> I think the `assert` is actually for making sure our two local copies in `streams` and `streamsByName` are the same.

I see. Then I guess here, and also where we handle `ChannelDeleteEvent`, it should be possible to do a targeted `assert` for that internal data-structure check, using nothing but the data we have locally. We can look up the channel in our `streams` map, get its name, look up the name in our `streamsByName` map, and check that the `ZulipStream` there matches the `ZulipStream` in `streams`.

And for the ways that unexpected server behavior can mislead our app, we can add checks that do run in production. Those ways, I think, are:

- The channel ID in the event isn't found in our data structures (our `streams` map). Let's early-return in this case, as in your PR, but add a `// TODO(log)`.
- The event's idea of the current channel name doesn't match our data structures. We can `// TODO(log)` in this case, but we can probably cheerfully continue handling the event, just ignoring the name in the event to keep our `streams` and `streamsByName` consistent with each other. Since the server is giving unexpected behavior, I think there's no strong reason to trust the name in the event more than the name in our data structures. But the `TODO(log)` could be helpful for catching bugs where our expectations are misaligned with server behavior.

Could you send a revision with a separate commit that does this for `ChannelDeleteEvent`, and also does it here for `ChannelUpdateEvent`?

> But if we really want to prevent getting `streamName` from the event, we can get it by storing `channel.name` in a temporary variable. Should I go with the latter option?

Yeah, this seems helpful. I've just checked, and the legacy app doesn't use the channel name on this event. Since it's possible to avoid using it now (just by using the name from the entry our `streams` map), let's avoid it, and that'll make it easy if the API wants to deprecate/remove the field in the future.